SimCalorimetry/HGCalSimProducers/src/HGCDigitizer.cc

```diff
@@ -114,75 +122,152 @@ void HGCDigitizer::accumulate(PileUpEventPrincipal const& e, edm::EventSetup con
+
+      //skip this hit if after ganging it is not valid
+      int layer(simId.layer()), cell(simId.cell());
+      std::pair<int,int> recoLayerCell=dddConst.simToReco(cell,layer,topo.detectorType());
+      cell  = recoLayerCell.first;
+      layer = recoLayerCell.second;
+      if(layer<0 || cell<0) 
+	{
+	  hitRefs[i]=std::make_tuple( i, 0, 0.);
+	  continue;
+	}
+
+      //assign the RECO DetId
+      DetId id( producesEEDigis() ?
+		(uint32_t)HGCEEDetId(mySubDet_,simId.zside(),layer,simId.sector(),simId.subsector(),cell):
+		(uint32_t)HGCHEDetId(mySubDet_,simId.zside(),layer,simId.sector(),simId.subsector(),cell) );
+
+      if (verbosity_>0) {
+	if (producesEEDigis())
+          edm::LogInfo("HGCDigitizer") <<" i/p " << simId << " o/p " << HGCEEDetId(id) << std::endl;
+	else
+          edm::LogInfo("HGCDigitizer") << " i/p " << simId << " o/p " << HGCHEDetId(id) << std::endl;
+      }
+
+      hitRefs[i]=std::make_tuple( i, 
+				  id.rawId(), 
+				  (float)hits->at(i).time() );
     }
+  std::sort(hitRefs.begin(),hitRefs.end(),this->orderByDetIdThenTime);
+  
+  //loop over sorted hits
+  for(int i=0; i<nchits; i++)
+    {
+      int hitidx   = std::get<0>(hitRefs[i]);
+      uint32_t id  = std::get<1>(hitRefs[i]);
+      if(id==0) continue; // to be ignored at RECO level
+
+      float toa    = std::get<2>(hitRefs[i]);
+      const PCaloHit &hit=hits->at( hitidx );     
+      float charge = hit.energy()*1e6*keV2fC;
 
-    //distance to the center of the detector
-    float dist2center( geom->getPosition(id).mag() );
 
-    //hit time: [time()]=ns  [centerDist]=cm [refSpeed_]=cm/ns + delay by 1ns
-    //accumulate in 6 buckets of 25ns (4 pre-samples, 1 in-time, 1 post-sample)
-    float tof(hit_it->time()-dist2center/refSpeed_+tofDelay_);
-    int itime=floor( tof/bxTime_ ) ;
+      //distance to the center of the detector
+      float dist2center( geom->getPosition(id).mag() );
       
-    //no need to add bx crossing - tof comes already corrected from the mixing module
-    //itime += bxCrossing;
-    itime += 4;
+      //hit time: [time()]=ns  [centerDist]=cm [refSpeed_]=cm/ns + delay by 1ns
+      //accumulate in 15 buckets of 25ns (9 pre-samples, 1 in-time, 5 post-samples)
+      float tof(toa-dist2center/refSpeed_+tofDelay_);
+      int itime=floor( tof/bxTime_ ) ;
       
-    if(itime<0 || itime>5) continue; 
+      //no need to add bx crossing - tof comes already corrected from the mixing module
+      //itime += bxCrossing;
+      itime += 9;
       
-    //energy deposited 
-    HGCSimEn_t ien( hit_it->energy() );
+      if(itime<0 || itime>14) continue; 
+            
+      //check if already existing (perhaps could remove this in the future - 2nd event should have all defined)
+      HGCSimHitDataAccumulator::iterator simHitIt=simHitAccumulator_->find(id);
+      if(simHitIt==simHitAccumulator_->end())
+	{
+	  simHitAccumulator_->insert( std::make_pair(id,baseData) );
+	  simHitIt=simHitAccumulator_->find(id);
```

Comment:
@ianna Yes, this one can be done better, fairly inefficient, know about proper use of insert. (didn't write this code)
Not actually a big time waster though, mostly saved by being a hash map rather than b-tree.